src/fsharp/service/ServiceAnalysis.fs

```diff
@@ -78,59 +81,62 @@ module UnusedOpens =
+                               Range = range
+                               AppliedScope = openDecl.AppliedScope }
+                | _ -> None)
 
     /// Only consider symbol uses which are the first part of a long ident, i.e. with no qualifying identifiers
-    let filterSymbolUses (getSourceLineStr: int -> string) (symbolUses: FSharpSymbolUse[]) : FSharpSymbolUse[] =
+    let filterSymbolUses (getSourceLineStr: int -> string) (symbolUses: seq<FSharpSymbolUse>) =
         symbolUses
-        |> Array.filter (fun su ->
-             match su.Symbol with
-             | :? FSharpMemberOrFunctionOrValue as fv when fv.IsExtensionMember -> 
+        |> Seq.filter(fun (su: FSharpSymbolUse) ->
+            match su.Symbol with
+            | :? FSharpMemberOrFunctionOrValue as fv when fv.IsExtensionMember -> 
                 // Extension members should be taken into account even though they have a prefix (as they do most of the time)
                 true
 
-             | :? FSharpMemberOrFunctionOrValue as fv when not fv.IsModuleValueOrMember -> 
+            | :? FSharpMemberOrFunctionOrValue as fv when not fv.IsModuleValueOrMember -> 
                 // Local values can be ignored
                 false
 
-             | :? FSharpMemberOrFunctionOrValue when su.IsFromDefinition -> 
+            | :? FSharpMemberOrFunctionOrValue when su.IsFromDefinition -> 
                 // Value definitions should be ignored
                 false
 
-             | :? FSharpGenericParameter -> 
+            | :? FSharpGenericParameter -> 
                 // Generic parameters can be ignored, they never come into scope via 'open'
                 false
 
-             | :? FSharpUnionCase when su.IsFromDefinition -> 
+            | :? FSharpUnionCase when su.IsFromDefinition -> 
                 false
 
-             | :? FSharpField as field when
-                     field.DeclaringEntity.IsSome && field.DeclaringEntity.Value.IsFSharpRecord ->
+            | :? FSharpField as field when
+                    field.DeclaringEntity.IsSome && field.DeclaringEntity.Value.IsFSharpRecord ->
                 // Record fields are used in name resolution
                 true
 
-             | :? FSharpField as field when field.IsUnionCaseField ->
-                 false
+            | :? FSharpField as field when field.IsUnionCaseField ->
+                false
 
-             | _ ->
+            | _ ->
                 // For the rest of symbols we pick only those which are the first part of a long ident, because it's they which are
                 // contained in opened namespaces / modules. For example, we pick `IO` from long ident `IO.File.OpenWrite` because
                 // it's `open System` which really brings it into scope.
                 let partialName = QuickParse.GetPartialLongNameEx (getSourceLineStr su.RangeAlternate.StartLine, su.RangeAlternate.EndColumn - 1)
                 List.isEmpty partialName.QualifyingIdents)
+        |> Array.ofSeq
```

Comment:
I guess ideally the consumption points where we consume these very long and expeensive sequences should always use a cancellation token, e.g. a new thing `|> Array.toSeqCancellable token` that drives the iterator in a loop and checks the cancellation at each step.

It normally doesn't matter but for these sequences I think it might?

Reply:
In this case, the operation isn't that time-consuming. It's the `for` loop that follows that is wild. That one gets all declaration items for each symbol as it processes them, which is the bulk of CPU time.

Doing a  `Array.toSeqCancellable token` could maybe speed things up but it wouldn't do much I think.